File: doc/codes/python/examples/tutorial/performance.py
import asyncio
import collections
import math

import numpy as np

import autd3_link_echocat as echocat
import autd3_pattern as pattern
from autd3 import MAX_INFLIGHT, Client, ClientConfig
from autd3.commands import ConfigPattern, SetSilencer, WritePatternBuffer
from autd3.geometry import Autd3, Geometry
from autd3.units import m, s
from autd3.value import LoopBehavior, PatternBank, SamplingConfig

NUM_POINTS = 1000
RADIUS_MM = 30.0


async def main() -> None:
    geometry = Geometry([Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client = await Client.open(
        geometry,
        echocat.EchocatLinkOption(),
        ClientConfig(),
    )

    patterns = geometry.pattern_buffer()

    # ANCHOR: configure
    builder = client.datagram_builder()
    builder.push(SetSilencer.disable())
    builder.push(
        WritePatternBuffer(
            bank=PatternBank.B0,
            index=0,
            emissions=patterns,
        )
    )
    builder.push(
        ConfigPattern(
            bank=PatternBank.B0,
            config=SamplingConfig.FREQ_40K,
            size=1,
            loop_behavior=LoopBehavior.Infinite,
        )
    )
    for frame in builder.build():
        await client.send_checked(frame)
    # ANCHOR_END: configure

    center = geometry.center() + np.array([0.0, 0.0, 150.0])
    wavelength = pattern.wavelength(340 * m / s)

    # ANCHOR: hot_loop
    pending = collections.deque()
    for i in range(NUM_POINTS):
        theta = 2.0 * math.pi * i / NUM_POINTS
        target = center + np.array([RADIUS_MM * math.cos(theta), RADIUS_MM * math.sin(theta), 0.0])
        pattern.focus(
            geometry,
            target,
            wavelength,
            pattern.FocusOption(),
            patterns,
        )
        builder = client.datagram_builder()
        builder.push(
            WritePatternBuffer(
                bank=PatternBank.B0,
                index=0,
                emissions=patterns,
            )
        )
        for frame in builder.build():
            if len(pending) >= MAX_INFLIGHT:
                (await pending.popleft()).check()
            pending.append(await client.send(frame))
    while pending:
        (await pending.popleft()).check()
    # ANCHOR_END: hot_loop

    await client.stop()
    await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: doc/codes/rust/examples/tutorial/performance.rs
use std::collections::VecDeque;
use std::f32::consts::PI;

use anyhow::Result;

use autd3_rs::commands::{ConfigPattern, SetSilencer, WritePatternBuffer};
use autd3_rs::geometry::{Autd3, Geometry, offset};
use autd3_rs::units::{m, mm, s};
use autd3_rs::value::{LoopBehavior, PatternBank, SamplingConfig};
use autd3_rs::{Client, ClientConfig, Frames, MAX_IN_FLIGHT, ResponseFuture};
use autd3_rs_link_nop::Nop;

const NUM_POINTS: usize = 1000;
const RADIUS_MM: f32 = 30.0;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(&geometry, Nop, ClientConfig::default()).await?;

    let mut patterns = geometry.pattern_buffer();

    {
        // ANCHOR: configure
        let mut builder = client.datagram_builder();
        builder
            .push(SetSilencer::disable())
            .push(WritePatternBuffer {
                bank: PatternBank::B0,
                index: 0,
                emissions: &patterns,
            })
            .push(ConfigPattern {
                bank: PatternBank::B0,
                config: SamplingConfig::FREQ_40K,
                size: 1,
                loop_behavior: LoopBehavior::Infinite,
            });
        for frame in &builder.build()? {
            client.send_checked(frame).await?;
        }
        // ANCHOR_END: configure
    }

    let center = geometry.center() + offset(0.0 * mm, 0.0 * mm, 150.0 * mm);
    let wavelength = autd3_rs_pattern::wavelength(340.0 * m / s);

    // ANCHOR: hot_loop
    let mut buf = Frames::default();
    let mut pending: VecDeque<ResponseFuture> = VecDeque::with_capacity(MAX_IN_FLIGHT);
    for i in 0..NUM_POINTS {
        let theta = 2.0 * PI * i as f32 / NUM_POINTS as f32;
        let target = center
            + offset(
                RADIUS_MM * theta.cos() * mm,
                RADIUS_MM * theta.sin() * mm,
                0.0 * mm,
            );
        autd3_rs_pattern::focus(
            &geometry,
            target,
            wavelength,
            &autd3_rs_pattern::FocusOption::default(),
            &mut patterns,
        );

        let mut builder = client.datagram_builder();
        builder.push(WritePatternBuffer {
            bank: PatternBank::B0,
            index: 0,
            emissions: &patterns,
        });
        builder.build_into(&mut buf)?;
        for frame in &buf {
            if pending.len() >= MAX_IN_FLIGHT {
                pending.pop_front().expect("non-empty").await?.check()?;
            }
            pending.push_back(client.send(frame).await?);
        }
    }
    while let Some(fut) = pending.pop_front() {
        fut.await?.check()?;
    }
    // ANCHOR_END: hot_loop

    client.stop().await?;
    client.close().await?;
    Ok(())
}
